AsyncDisplayKit/Details/ASCollectionViewLayoutController.mm

```diff
@@ -0,0 +1,153 @@
+                                                  rangeTuningParameters.trailingBufferScreenfuls;
+  verticalBuffer.negativeDirection = movingDown ? rangeTuningParameters.trailingBufferScreenfuls :
+                                                  rangeTuningParameters.leadingBufferScreenfuls;
+  return verticalBuffer;
+}
+
+struct ASRangeGeometry {
+  CGRect rangeBounds;
+  CGRect updateBounds;
+};
+typedef struct ASRangeGeometry ASRangeGeometry;
+
+
+#pragma mark -
+#pragma mark ASCollectionViewLayoutController
+
+@interface ASCollectionViewLayoutController () {
+  ASCollectionView * __weak _collectionView;
+  std::vector<CGRect> _updateRangeBoundsIndexedByRangeType;
+}
+@end
+
+@implementation ASCollectionViewLayoutController
+
+- (instancetype)initWithCollectionView:(ASCollectionView *)collectionView {
+  if (!(self = [super init])) {
+    return nil;
+  }
+  _collectionView = collectionView;
+  _updateRangeBoundsIndexedByRangeType = std::vector<CGRect>(ASLayoutRangeTypeCount);
+  return self;
+}
+
+#pragma mark -
+#pragma mark Index Paths in Range
+
+- (NSSet *)indexPathsForScrolling:(ASScrollDirection)scrollDirection
+                     viewportSize:(CGSize)viewportSize
+                        rangeType:(ASLayoutRangeType)rangeType {
+  ASRangeGeometry rangeGeometry = [self rangeGeometryWithScrollDirection:scrollDirection
+                                                          collectionView:_collectionView
+                                                   rangeTuningParameters:[self tuningParametersForRangeType:rangeType]];
+  _updateRangeBoundsIndexedByRangeType[rangeType] = rangeGeometry.updateBounds;
+  return [self indexPathsForItemsWithinRangeBounds:rangeGeometry.rangeBounds collectionView:_collectionView];
+}
+
+- (ASRangeGeometry)rangeGeometryWithScrollDirection:(ASScrollDirection)scrollDirection
+                                     collectionView:(ASCollectionView *)collectionView
+                              rangeTuningParameters:(ASRangeTuningParameters)rangeTuningParameters {
+  CGRect rangeBounds = collectionView.bounds;
+  CGRect updateBounds = collectionView.bounds;
+  ASScrollDirection scrollableDirections = [collectionView scrollableDirections];
+  
+  BOOL canScrollHorizontally = ASScrollDirectionContainsHorizontalDirection(scrollableDirections);
+  if (canScrollHorizontally) {
+    ASDirectionalScreenfulBuffer horizontalBuffer = ASDirectionalScreenfulBufferHorizontal(scrollDirection,
+                                                                                           rangeTuningParameters);
+    rangeBounds = asdk_CGRectExpandHorizontally(rangeBounds,
+                                                horizontalBuffer.negativeDirection,
+                                                horizontalBuffer.positiveDirection);
+    // Update bounds is at most 95% of the next/previous screenful and at least half of tuning parameter value.
```

Comment:
I don't completely understand this yet.  Why 95%?  Can this be better-documented?